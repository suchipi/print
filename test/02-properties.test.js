const { print } = require("./helpers");

describe("Property fields", () => {
	it("prints string-valued properties", () => {
		expect(print({ foo: "bar" })).toMatchSnapshot();
		expect(print({ 123: "baz" })).toMatchSnapshot();
		expect(print({ baz: 123.4 })).toMatchSnapshot();
	});

	it("prints symbol-valued properties", () => {
		expect(print({ [Symbol("foo")]: "bar" })).toMatchSnapshot();
		expect(print({ [Symbol.toStringTag]: "foo" })).toMatchSnapshot();
		expect(
			print({ [Symbol.toStringTag]: "foo" }, { noAmp: true })
		).toMatchSnapshot();
		expect(print({ [Symbol.iterator]: Symbol.iterator })).toMatchSnapshot();
		expect(
			print({ [Symbol.iterator]: Symbol.iterator }, { noAmp: true })
		).toMatchSnapshot();
	});

	it("prints nested properties", () => {
		expect(print({ foo: { bar: "baz" } })).toMatchSnapshot();
		expect(print({ foo: { bar: { baz: "qux" } } })).toMatchSnapshot();
	});

	it("prints empty objects on one line", () => {
		expect(print({})).toMatchSnapshot();
		expect(print({ foo: {} })).toMatchSnapshot();
	});

	describe("Extensibility", () => {
		it("identifies non-extensible objects", () => {
			expect(print(Object.preventExtensions({ foo: 1 }))).toMatchSnapshot();
		});

		it("identifies sealed objects", () => {
			expect(print(Object.seal({ foo: 1 }))).toMatchSnapshot();
		});

		it("identifies frozen objects", () => {
			expect(print(Object.freeze({}))).toMatchSnapshot();
			expect(print(Object.freeze({ foo: 1 }))).toMatchSnapshot();
		});
	});

	describe("Prototypes", () => {
		it("identifies null prototypes", () => {
			const obj = { __proto__: null };
			expect(print(obj)).toMatchSnapshot();
			obj.foo = "Foo";
			expect(print(obj)).toMatchSnapshot();
			obj.bar = "Bar";
			expect(print(obj)).toMatchSnapshot();
			expect(print(Object.freeze(obj))).toMatchSnapshot();
		});

		it("prints `__proto__` if `opts.proto` is enabled", () => {
			const foo = { name: "Foo" };
			const bar = { name: "Bar" };
			bar.__proto__ = foo;
			expect(print({ foo, bar }, { proto: true })).toMatchSnapshot();
		});

		it("recovers gracefully if `__proto__` access throws", () => {
			const foo = { name: "Foo" };
			Object.defineProperty(foo, "__proto__", {
				get() {
					const error = new Error("Don't.");
					error.no = "Really";
					throw error;
				},
			});
			expect(print({ foo }, { proto: true })).toMatchSnapshot();
		});
	});

	describe("“Magic” numbers", () => {
		it("identifies Math.* constants", () => {
			expect(
				print({
					euler: Math.E,
					obj: {
						pi: Math.PI,
					},
				})
			).toMatchSnapshot();
		});

		it("identifies Number.* constants", () => {
			expect(
				print({
					epsilon: Number.EPSILON,
					range: {
						min: Number.MIN_VALUE,
						max: Number.MAX_VALUE,
					},
				})
			).toMatchSnapshot();
		});

		it("doesn't identify them when printing their owners", () => {
			let lines = print(Math, { all: true }).split("\n");
			expect(lines.some((line) => "PI: " + Math.PI === line.trim())).toBe(true);
			expect(lines.some((line) => line.trim() === "PI: Math.PI")).toBe(false);

			lines = print(Number, { all: true }).split("\n");
			expect(
				lines.some((line) => "MAX_VALUE: " + Number.MAX_VALUE === line.trim())
			).toBe(true);
			expect(
				lines.some((line) => line.trim() === "MAX_VALUE: Number.MAX_VALUE")
			).toBe(false);

			const name = "REFERENCE_TO_MAGICAL_NUMBER_THING";
			Math[name] = Number.MAX_VALUE;
			lines = print(Math, { all: true }).split("\n");
			expect(
				lines.some((line) => line.trim() === name + ": Number.MAX_VALUE")
			).toBe(true);

			delete Math[name];
			Number[name] = Math.PI;
			lines = print(Number, { all: true }).split("\n");
			expect(lines.some((line) => line.trim() === name + ": Math.PI")).toBe(
				true
			);
			delete Number[name];
		});
	});

	describe("Ordering", () => {
		const Γ = Symbol("gamma");
		const Ζ = Symbol("zeta");
		const Β = Symbol("beta");
		const Τ = Symbol("tau");
		const Α = Symbol("alpha");

		describe("Default", () => {
			it("sorts names in creation order", () => {
				expect(
					print({
						G: "gamma",
						Z: "zeta",
						B: "beta",
						T: "tau",
						A: "alpha",
					})
				).toMatchSnapshot();
				const obj = { foo: "Foo" };
				obj.bar = "Bar";
				expect(print(obj)).toMatchSnapshot();
				delete obj.foo;
				obj.xyz = "XYZ";
				obj.abc = "ABC";
				obj.foo = "Foo";
				expect(print(obj)).toMatchSnapshot();
			});

			it("sorts symbols in creation order", () => {
				expect(
					print({
						[Γ]: "G",
						[Ζ]: "Z",
						[Β]: "B",
						[Τ]: "T",
						[Α]: "A",
					})
				).toMatchSnapshot();
				const foo = Symbol("foo");
				const obj = { [foo]: "Foo" };
				obj[Symbol("bar")] = "Bar";
				expect(print(obj)).toMatchSnapshot();
				delete obj[foo];
				obj[Symbol("xyz")] = "XYZ";
				obj[Symbol("abc")] = "ABC";
				obj[foo] = "Foo";
				expect(print(obj)).toMatchSnapshot();
			});

			it("sorts names and symbols separately", () => {
				const foo = Symbol("foo");
				const bar = Symbol("bar");
				const qux = Symbol("qux");
				const obj = {
					xyz: "XYZ",
					[foo]: "Foo",
					abc: "ABC",
					[bar]: "Bar",
				};
				expect(print(obj)).toMatchSnapshot();
				obj.foo = "Foo";
				obj[qux] = "Qux";
				expect(print(obj)).toMatchSnapshot();
				delete obj.xyz;
				obj.xyz = "XYZ";
				expect(print(obj)).toMatchSnapshot();
			});
		});

		describe("When `sort` is enabled", () => {
			it("sorts names alphabetically", () => {
				const obj = {
					G: "gamma",
					D: "delta",
					A: "alpha",
					B: "beta",
					P: "pi",
					__FB: "FooBar",
				};
				expect(print(obj, { sort: true })).toMatchSnapshot();
				obj.Z = "zeta";
				obj.E = "epsilon";
				expect(print(obj, { sort: true })).toMatchSnapshot();
			});

			it("sorts symbols alphabetically", () => {
				const obj = {
					[Γ]: "G",
					[Β]: "B",
					[Τ]: "T",
				};
				expect(print(obj, { sort: true })).toMatchSnapshot();
				obj[Ζ] = "Z";
				obj[Α] = "A";
				expect(print(obj, { sort: true })).toMatchSnapshot();
			});

			it("sorts symbols and names together", () => {
				const obj = {
					xyz: "XYZ",
					[Symbol("uvw")]: "UVW",
					abc: "ABC",
					[Symbol("def")]: "DEF",
				};
				expect(print(obj, { sort: true })).toMatchSnapshot();
			});

			it("sorts case-insensitively", () => {
				expect(
					print(
						{
							D: 0,
							c: 3,
							a: 1,
							B: 2,
						},

						{ sort: true }
					)
				).toMatchSnapshot();
				expect(
					print(
						{
							[Symbol("D")]: 0,
							[Symbol("c")]: 3,
							[Symbol("a")]: 1,
							[Symbol("B")]: 2,
						},

						{ sort: true }
					)
				).toMatchSnapshot();
			});
		});
	});

	describe("Visibility", () => {
		it("hides non-enumerable properties", () => {
			const obj = {
				c: "C",
				b: "B",
				a: "A",
			};
			Object.defineProperty(obj, "b", { enumerable: false });
			expect(print(obj)).toMatchSnapshot();
			expect(print(obj, { sort: true })).toMatchSnapshot();
		});

		it("shows them if `opts.all` is enabled", () => {
			const obj = Object.defineProperties(
				{},
				{
					abc: { enumerable: false, value: "ABC" },
					xyz: { enumerable: true, value: "XYZ" },
				}
			);
			expect(print(obj)).toMatchSnapshot();
			expect(print(obj, { all: true })).toMatchSnapshot();
		});

		it("never shows inherited properties", () => {
			const descriptors = {
				foo: { value: "Foo", writable: true, enumerable: true },
				bar: { value: "Bar", writable: true, enumerable: false },
			};
			class Thing {
				method() {}
			}
			Object.defineProperties(Thing.prototype, descriptors);
			const value = new Thing();
			expect(print(value)).toMatchSnapshot();
			expect(print(value, { all: true })).toMatchSnapshot();
			Object.defineProperties(value, descriptors);
			expect(print(value)).toMatchSnapshot();
			expect(print(value, { all: true })).toMatchSnapshot();
		});
	});

	describe("Accessors", () => {
		it("doesn't invoke getter functions", () => {
			let called = false;
			const obj = {
				get bar() {
					called = true;
					return "Bar";
				},
			};
			print(obj);
			expect(called).toBe(false);
			expect(obj.bar).toEqual("Bar");
			expect(called).toBe(true);
		});

		it("invokes them if `followGetters` is enabled", () => {
			let callCount = 0;
			expect(
				print(
					{
						foo: "Foo",
						get bar() {
							return ++callCount;
						},
						baz: "Baz",
					},

					{ followGetters: true }
				)
			).toMatchSnapshot();
			expect(callCount).toEqual(1);
		});

		it("invokes non-enumerable getters only if `all` is enabled", () => {
			let abc = 0;
			let xyz = 0;
			const obj = Object.defineProperties(
				{},
				{
					abc: {
						enumerable: false,
						get() {
							return ++abc;
						},
					},
					xyz: {
						enumerable: true,
						get() {
							return ++xyz;
						},
					},
				}
			);
			expect(print(obj, { followGetters: true })).toMatchSnapshot();
			expect(abc).toEqual(0);
			expect(xyz).toEqual(1);
			expect(print(obj, { followGetters: true, all: true })).toMatchSnapshot();
			expect(abc).toEqual(1);
			expect(xyz).toEqual(2);
		});

		it("catches and prints any error that's thrown", () => {
			const obj = {
				abc: "ABC",
				get foo() {
					const error = new Error("Don't touch me");
					error.y = "tho";
					throw error;
				},
				xyz: "XYZ",
			};
			expect(print(obj, { followGetters: true })).toMatchSnapshot();
		});
	});

	describe("When `maxDepth` is exceeded", () => {
		it("elides property lists", () => {
			const obj = { foo: { bar: { baz: { qux: { qul: 1 } } } } };
			expect(print(obj, { maxDepth: 0 })).toMatchSnapshot();
			expect(print(obj, { maxDepth: -1 })).toMatchSnapshot();
			expect(print(obj, { maxDepth: 1 })).toMatchSnapshot();
			expect(print(obj, { maxDepth: 2 })).toMatchSnapshot();
			expect(print(obj, { maxDepth: 3 })).toMatchSnapshot();
			expect(print(obj, { maxDepth: 4 })).toMatchSnapshot();
			expect(print(obj, { maxDepth: 5 })).toMatchSnapshot();
			expect(print(obj, { maxDepth: NaN })).toMatchSnapshot();
			expect(print(obj, { maxDepth: Infinity })).toMatchSnapshot();
		});

		it("elides details about null prototypes", () => {
			expect(print({ __proto__: null }, { maxDepth: 0 })).toMatchSnapshot();
			expect(print({ __proto__: null }, { maxDepth: 1 })).toMatchSnapshot();
			expect(
				print({ foo: { __proto__: null } }, { maxDepth: 1 })
			).toMatchSnapshot();
			expect(
				print({ foo: { __proto__: null } }, { maxDepth: 2 })
			).toMatchSnapshot();
		});

		it("elides details about extensibility", () => {
			expect(print(Object.freeze({}), { maxDepth: 0 })).toMatchSnapshot();
			expect(print(Object.freeze({}), { maxDepth: 1 })).toMatchSnapshot();
			expect(
				print({ foo: Object.freeze({}) }, { maxDepth: 1 })
			).toMatchSnapshot();
			expect(
				print({ foo: Object.freeze({}) }, { maxDepth: 2 })
			).toMatchSnapshot();
			expect(
				print(Object.freeze({ __proto__: null }), {
					maxDepth: 0,
				})
			).toMatchSnapshot();
			expect(
				print(Object.freeze({ __proto__: null }), { maxDepth: 1 })
			).toMatchSnapshot();
			expect(
				print({ foo: Object.freeze({ __proto__: null }) }, { maxDepth: 1 })
			).toMatchSnapshot();
			expect(
				print({ foo: Object.freeze({ __proto__: null }) }, { maxDepth: 2 })
			).toMatchSnapshot();

			expect(
				print({ __proto__: null, foo: Object.freeze({}) }, { maxDepth: 1 })
			).toMatchSnapshot();
			expect(
				print({ __proto__: null, foo: Object.freeze({}) }, { maxDepth: 2 })
			).toMatchSnapshot();
		});

		it("doesn't needlessly invoke getters", () => {
			let callCount = 0;
			const foo = {
				get bar() {
					return ++callCount;
				},
			};
			expect(
				print({ foo }, { followGetters: true, maxDepth: 1 })
			).toMatchSnapshot();
			expect(callCount).toEqual(0);
			expect(
				print({ foo }, { followGetters: true, maxDepth: 2 })
			).toMatchSnapshot();
			expect(callCount).toEqual(1);
			expect(
				print(
					{
						get foo() {
							return foo;
						},
					},

					{ followGetters: true, maxDepth: 2 }
				)
			).toMatchSnapshot();
			expect(callCount).toEqual(2);
		});

		it("still prints the object's class", () => {
			class Thing {
				constructor(name) {
					this.name = name;
				}
			}
			expect(print(new Thing("Foo"), { maxDepth: 0 })).toMatchSnapshot();
			expect(print(new Thing("Foo"), { maxDepth: 1 })).toMatchSnapshot();
			expect(
				print({ foo: new Thing("Bar") }, { maxDepth: 1 })
			).toMatchSnapshot();
			expect(
				print({ foo: new Thing("Bar") }, { maxDepth: 2 })
			).toMatchSnapshot();
		});

		it("still prints primitive values", () => {
			expect(print("Foo", { maxDepth: 0 })).toMatchSnapshot();
			expect(print(3.525, { maxDepth: 0 })).toMatchSnapshot();
			expect(print(false, { maxDepth: 0 })).toMatchSnapshot();
			expect(print(Symbol("Foo"), { maxDepth: 0 })).toMatchSnapshot();
			expect(
				print({ foo: "Foo", bar: 1, baz: false }, { maxDepth: 1 })
			).toMatchSnapshot();
			expect(
				print({ [Symbol("Foo")]: "Bar" }, { maxDepth: 1 })
			).toMatchSnapshot();
			expect(print({ [Symbol("Foo")]: {} }, { maxDepth: 1 })).toMatchSnapshot();
		});
	});
});
