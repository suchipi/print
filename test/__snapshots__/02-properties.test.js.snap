// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property fields Accessors catches and prints any error that's thrown 1`] = `
{
	abc: "ABC"
	foo: Error {
		Error: Don't touch me
			at Object.get foo [as foo] (<rootDir>/test/02-properties.test.js:LINE:COL)
			at print (<rootDir>/print.js:LINE:COL)
			at realPrint (<rootDir>/test/helpers.js:LINE:COL)
			at Object.print (<rootDir>/test/02-properties.test.js:LINE:COL)
			at Promise.then.completed (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
			at new Promise (<anonymous>)
			at callAsyncCircusFn (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
			at _callCircusTest (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
			at processTicksAndRejections (node:internal/process/task_queues:LINE:COL)
			at _runTest (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
		
		y: "tho"
	}
	xyz: "XYZ"
}
`;

exports[`Property fields Accessors invokes non-enumerable getters only if \`all\` is enabled 1`] = `
{
	xyz: 1
}
`;

exports[`Property fields Accessors invokes non-enumerable getters only if \`all\` is enabled 2`] = `
{
	abc: 1
	xyz: 2
}
`;

exports[`Property fields Accessors invokes them if \`followGetters\` is enabled 1`] = `
{
	foo: "Foo"
	bar: 1
	baz: "Baz"
}
`;

exports[`Property fields Extensibility identifies frozen objects 1`] = `
{
	Frozen
}
`;

exports[`Property fields Extensibility identifies frozen objects 2`] = `
{
	Frozen
	
	foo: 1
}
`;

exports[`Property fields Extensibility identifies non-extensible objects 1`] = `
{
	Non-extensible
	
	foo: 1
}
`;

exports[`Property fields Extensibility identifies sealed objects 1`] = `
{
	Sealed
	
	foo: 1
}
`;

exports[`Property fields Ordering Default sorts names and symbols separately 1`] = `
{
	xyz: "XYZ"
	abc: "ABC"
	Symbol(foo): "Foo"
	Symbol(bar): "Bar"
}
`;

exports[`Property fields Ordering Default sorts names and symbols separately 2`] = `
{
	xyz: "XYZ"
	abc: "ABC"
	foo: "Foo"
	Symbol(foo): "Foo"
	Symbol(bar): "Bar"
	Symbol(qux): "Qux"
}
`;

exports[`Property fields Ordering Default sorts names and symbols separately 3`] = `
{
	abc: "ABC"
	foo: "Foo"
	xyz: "XYZ"
	Symbol(foo): "Foo"
	Symbol(bar): "Bar"
	Symbol(qux): "Qux"
}
`;

exports[`Property fields Ordering Default sorts names in creation order 1`] = `
{
	G: "gamma"
	Z: "zeta"
	B: "beta"
	T: "tau"
	A: "alpha"
}
`;

exports[`Property fields Ordering Default sorts names in creation order 2`] = `
{
	foo: "Foo"
	bar: "Bar"
}
`;

exports[`Property fields Ordering Default sorts names in creation order 3`] = `
{
	bar: "Bar"
	xyz: "XYZ"
	abc: "ABC"
	foo: "Foo"
}
`;

exports[`Property fields Ordering Default sorts symbols in creation order 1`] = `
{
	Symbol(gamma): "G"
	Symbol(zeta): "Z"
	Symbol(beta): "B"
	Symbol(tau): "T"
	Symbol(alpha): "A"
}
`;

exports[`Property fields Ordering Default sorts symbols in creation order 2`] = `
{
	Symbol(foo): "Foo"
	Symbol(bar): "Bar"
}
`;

exports[`Property fields Ordering Default sorts symbols in creation order 3`] = `
{
	Symbol(bar): "Bar"
	Symbol(xyz): "XYZ"
	Symbol(abc): "ABC"
	Symbol(foo): "Foo"
}
`;

exports[`Property fields Ordering When \`sort\` is enabled sorts case-insensitively 1`] = `
{
	a: 1
	B: 2
	c: 3
	D: 0
}
`;

exports[`Property fields Ordering When \`sort\` is enabled sorts case-insensitively 2`] = `
{
	Symbol(a): 1
	Symbol(B): 2
	Symbol(c): 3
	Symbol(D): 0
}
`;

exports[`Property fields Ordering When \`sort\` is enabled sorts names alphabetically 1`] = `
{
	__FB: "FooBar"
	A: "alpha"
	B: "beta"
	D: "delta"
	G: "gamma"
	P: "pi"
}
`;

exports[`Property fields Ordering When \`sort\` is enabled sorts names alphabetically 2`] = `
{
	__FB: "FooBar"
	A: "alpha"
	B: "beta"
	D: "delta"
	E: "epsilon"
	G: "gamma"
	P: "pi"
	Z: "zeta"
}
`;

exports[`Property fields Ordering When \`sort\` is enabled sorts symbols alphabetically 1`] = `
{
	Symbol(beta): "B"
	Symbol(gamma): "G"
	Symbol(tau): "T"
}
`;

exports[`Property fields Ordering When \`sort\` is enabled sorts symbols alphabetically 2`] = `
{
	Symbol(alpha): "A"
	Symbol(beta): "B"
	Symbol(gamma): "G"
	Symbol(tau): "T"
	Symbol(zeta): "Z"
}
`;

exports[`Property fields Ordering When \`sort\` is enabled sorts symbols and names together 1`] = `
{
	abc: "ABC"
	Symbol(def): "DEF"
	Symbol(uvw): "UVW"
	xyz: "XYZ"
}
`;

exports[`Property fields Prototypes identifies null prototypes 1`] = `
{
	Null prototype
}
`;

exports[`Property fields Prototypes identifies null prototypes 2`] = `
{
	Null prototype
	
	foo: "Foo"
}
`;

exports[`Property fields Prototypes identifies null prototypes 3`] = `
{
	Null prototype
	
	foo: "Foo"
	bar: "Bar"
}
`;

exports[`Property fields Prototypes identifies null prototypes 4`] = `
{
	Null prototype
	Frozen
	
	foo: "Foo"
	bar: "Bar"
}
`;

exports[`Property fields Prototypes prints \`__proto__\` if \`opts.proto\` is enabled 1`] = `
{
	__proto__: {
		Null prototype
		
		__proto__: null
	}
	foo: {
		__proto__: -> {root}.__proto__
		name: "Foo"
	}
	bar: {
		__proto__: -> {root}.foo
		name: "Bar"
	}
}
`;

exports[`Property fields Prototypes recovers gracefully if \`__proto__\` access throws 1`] = `
{
	__proto__: {
		Null prototype
		
		__proto__: null
	}
	foo: {
		__proto__: Error {
			Error: Don't.
				at Object.get (<rootDir>/test/02-properties.test.js:LINE:COL)
				at __proto__ (<rootDir>/print.js:LINE:COL)
				at print (<rootDir>/print.js:LINE:COL)
				at realPrint (<rootDir>/test/helpers.js:LINE:COL)
				at Object.print (<rootDir>/test/02-properties.test.js:LINE:COL)
				at Promise.then.completed (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
				at new Promise (<anonymous>)
				at callAsyncCircusFn (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
				at _callCircusTest (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
				at processTicksAndRejections (node:internal/process/task_queues:LINE:COL)
			
			no: "Really"
		}
		name: "Foo"
	}
}
`;

exports[`Property fields Visibility hides non-enumerable properties 1`] = `
{
	c: "C"
	a: "A"
}
`;

exports[`Property fields Visibility hides non-enumerable properties 2`] = `
{
	a: "A"
	c: "C"
}
`;

exports[`Property fields Visibility never shows inherited properties 1`] = `Thing {}`;

exports[`Property fields Visibility never shows inherited properties 2`] = `Thing {}`;

exports[`Property fields Visibility never shows inherited properties 3`] = `
Thing {
	foo: "Foo"
}
`;

exports[`Property fields Visibility never shows inherited properties 4`] = `
Thing {
	foo: "Foo"
	bar: "Bar"
}
`;

exports[`Property fields Visibility shows them if \`opts.all\` is enabled 1`] = `
{
	xyz: "XYZ"
}
`;

exports[`Property fields Visibility shows them if \`opts.all\` is enabled 2`] = `
{
	abc: "ABC"
	xyz: "XYZ"
}
`;

exports[`Property fields When \`maxDepth\` is exceeded doesn't needlessly invoke getters 1`] = `
{
	foo: {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded doesn't needlessly invoke getters 2`] = `
{
	foo: {
		bar: 1
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded doesn't needlessly invoke getters 3`] = `
{
	foo: {
		bar: 2
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 1`] = `{…}`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 2`] = `
{
	Frozen
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 3`] = `
{
	foo: {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 4`] = `
{
	foo: {
		Frozen
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 5`] = `{…}`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 6`] = `
{
	Null prototype
	Frozen
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 7`] = `
{
	foo: {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 8`] = `
{
	foo: {
		Null prototype
		Frozen
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 9`] = `
{
	Null prototype
	
	foo: {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about extensibility 10`] = `
{
	Null prototype
	
	foo: {
		Frozen
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about null prototypes 1`] = `{…}`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about null prototypes 2`] = `
{
	Null prototype
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about null prototypes 3`] = `
{
	foo: {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides details about null prototypes 4`] = `
{
	foo: {
		Null prototype
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 1`] = `{…}`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 2`] = `{…}`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 3`] = `
{
	foo: {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 4`] = `
{
	foo: {
		bar: {…}
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 5`] = `
{
	foo: {
		bar: {
			baz: {…}
		}
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 6`] = `
{
	foo: {
		bar: {
			baz: {
				qux: {…}
			}
		}
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 7`] = `
{
	foo: {
		bar: {
			baz: {
				qux: {
					qul: 1
				}
			}
		}
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 8`] = `
{
	foo: {
		bar: {
			baz: {
				qux: {
					qul: 1
				}
			}
		}
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded elides property lists 9`] = `
{
	foo: {
		bar: {
			baz: {
				qux: {
					qul: 1
				}
			}
		}
	}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded still prints primitive values 1`] = `"Foo"`;

exports[`Property fields When \`maxDepth\` is exceeded still prints primitive values 2`] = `3.525`;

exports[`Property fields When \`maxDepth\` is exceeded still prints primitive values 3`] = `false`;

exports[`Property fields When \`maxDepth\` is exceeded still prints primitive values 4`] = `Symbol(Foo)`;

exports[`Property fields When \`maxDepth\` is exceeded still prints primitive values 5`] = `
{
	foo: "Foo"
	bar: 1
	baz: false
}
`;

exports[`Property fields When \`maxDepth\` is exceeded still prints primitive values 6`] = `
{
	Symbol(Foo): "Bar"
}
`;

exports[`Property fields When \`maxDepth\` is exceeded still prints primitive values 7`] = `
{
	Symbol(Foo): {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded still prints the object's class 1`] = `Thing {…}`;

exports[`Property fields When \`maxDepth\` is exceeded still prints the object's class 2`] = `
Thing {
	name: "Foo"
}
`;

exports[`Property fields When \`maxDepth\` is exceeded still prints the object's class 3`] = `
{
	foo: Thing {…}
}
`;

exports[`Property fields When \`maxDepth\` is exceeded still prints the object's class 4`] = `
{
	foo: Thing {
		name: "Bar"
	}
}
`;

exports[`Property fields prints empty objects on one line 1`] = `{}`;

exports[`Property fields prints empty objects on one line 2`] = `
{
	foo: {}
}
`;

exports[`Property fields prints nested properties 1`] = `
{
	foo: {
		bar: "baz"
	}
}
`;

exports[`Property fields prints nested properties 2`] = `
{
	foo: {
		bar: {
			baz: "qux"
		}
	}
}
`;

exports[`Property fields prints string-valued properties 1`] = `
{
	foo: "bar"
}
`;

exports[`Property fields prints string-valued properties 2`] = `
{
	123: "baz"
}
`;

exports[`Property fields prints string-valued properties 3`] = `
{
	baz: 123.4
}
`;

exports[`Property fields prints symbol-valued properties 1`] = `
{
	Symbol(foo): "bar"
}
`;

exports[`Property fields prints symbol-valued properties 2`] = `
{
	@@toStringTag: "foo"
}
`;

exports[`Property fields prints symbol-valued properties 3`] = `
{
	Symbol(Symbol.toStringTag): "foo"
}
`;

exports[`Property fields prints symbol-valued properties 4`] = `
{
	@@iterator: @@iterator
}
`;

exports[`Property fields prints symbol-valued properties 5`] = `
{
	Symbol(Symbol.iterator): Symbol(Symbol.iterator)
}
`;

exports[`Property fields “Magic” numbers identifies Math.* constants 1`] = `
{
	euler: Math.E
	obj: {
		pi: Math.PI
	}
}
`;

exports[`Property fields “Magic” numbers identifies Number.* constants 1`] = `
{
	epsilon: Number.EPSILON
	range: {
		min: Number.MIN_VALUE
		max: Number.MAX_VALUE
	}
}
`;
